File: java/FlattenNestedListIterator.java
import java.util.*;

// https://leetcode.com/problems/flatten-nested-list-iterator/
/**
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 * public interface NestedInteger {
 *     // @return true if this NestedInteger holds a single integer, rather than a nested list.
 *     public boolean isInteger();
 *     // @return the single integer that this NestedInteger holds, if it holds a single integer
 *     // Return null if this NestedInteger holds a nested list
 *     public Integer getInteger();
 *     // @return the nested list that this NestedInteger holds, if it holds a nested list
 *     // Return empty list if this NestedInteger holds a single integer
 *     public List<NestedInteger> getList();
 * }
 */

interface NestedInteger {
    // @return true if this NestedInteger holds a single integer, rather than a nested list.
    public boolean isInteger();
    // @return the single integer that this NestedInteger holds, if it holds a single integer
    // Return null if this NestedInteger holds a nested list
    public Integer getInteger();
    // @return the nested list that this NestedInteger holds, if it holds a nested list
    // Return empty list if this NestedInteger holds a single integer
    public List<NestedInteger> getList();
}


class SingleIterator implements Iterator<Integer> {
    private final int value;
    private boolean iterated;

    SingleIterator(int value) {
        this.value = value;
    }

    @Override
    public boolean hasNext() {
        return !iterated;
    }

    @Override
    public Integer next() {
        iterated = true;
        return value;
    }
}

class NestedIterator implements Iterator<Integer> {
    private final Deque<Iterator<Integer>> stack = new LinkedList<>();

    public NestedIterator(List<NestedInteger> nestedList) {
        initFromList(nestedList);
    }

    NestedIterator(NestedInteger nestedInteger) {
        if (nestedInteger.isInteger()) {
            stack.push(new SingleIterator(nestedInteger.getInteger()));
        } else {
            initFromList(nestedInteger.getList());
        }
    }

    @Override
    public Integer next() {
        while (!stack.isEmpty() && !stack.peek().hasNext()) {
            stack.pop();
        }
        if (!hasNext() || stack.isEmpty()) {
            throw new IllegalStateException("No more items in iterator");
        }
        return stack.peek().next();
    }

    @Override
    public boolean hasNext() {
        while (!stack.isEmpty() && !stack.peek().hasNext()) {
            stack.pop();
        }
        return !stack.isEmpty() && stack.peek().hasNext();
    }

    private void initFromList(List<NestedInteger> nestedList) {
        var reverseIter = nestedList.listIterator(nestedList.size() - 1);
        while (reverseIter.hasPrevious()) {
            stack.push(new NestedIterator(reverseIter.previous()));
        }
    }
}

public class FlattenNestedListIterator {
    public static void main(String[] args) {
        System.out.println(
                fromIterator(
                        new NestedIterator(

                        )
                )
        );
    }

    static List<Integer> fromIterator(Iterator<Integer> iterator) {
        var arrayList = new ArrayList<Integer>();
        iterator.forEachRemaining(arrayList::add);
        return arrayList;
    }
}

/*
 * Your NestedIterator object will be instantiated and called as such:
 * NestedIterator i = new NestedIterator(nestedList);
 * while (i.hasNext()) v[f()] = i.next();
 */

File: typescript/flatten-nested-list-iterator.ts
// https://leetcode.com/problems/flatten-nested-list-iterator/
/**
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 * class NestedInteger {
 *     If value is provided, then it holds a single integer
 *     Otherwise it holds an empty nested list
 *     constructor(value?: number) {
 *         ...
 *     };
 *
 *     Return true if this NestedInteger holds a single integer, rather than a nested list.
 *     isInteger(): boolean {
 *         ...
 *     };
 *
 *     Return the single integer that this NestedInteger holds, if it holds a single integer
 *     Return null if this NestedInteger holds a nested list
 *     getInteger(): number | null {
 *         ...
 *     };
 *
 *     Set this NestedInteger to hold a single integer equal to value.
 *     setInteger(value: number) {
 *         ...
 *     };
 *
 *     Set this NestedInteger to hold a nested list and adds a nested integer elem to it.
 *     add(elem: NestedInteger) {
 *         ...
 *     };
 *
 *     Return the nested list that this NestedInteger holds,
 *     or an empty list if this NestedInteger holds a single integer
 *     getList(): NestedInteger[] {
 *         ...
 *     };
 * };
 */
export class NestedInteger {
  private value: number | null;
  private values: NestedInteger[] = [];

  constructor(value: number | null = null) {
    this.value = value;
  }

  isInteger(): boolean {
    return this.value !== null;
  }

  getInteger(): number | null {
    return this.value ?? null;
  }
  setInteger(value: number) {
    this.value = value;
  }
  add(elem: NestedInteger) {
    if (this.isInteger()) {
      throw new Error("Cannot add items to integer");
    }
    this.values.push(elem);
  }
  getList(): NestedInteger[] {
    if (this.isInteger()) {
      throw new Error("Cannot transform integer to list");
    }
    return this.values;
  }
}

export interface ListIterator<T> {
  hasNext(): boolean;
  next(): T;
}

export class Single<T> implements ListIterator<T> {
  private value: T;
  private iterated: boolean;

  constructor(value: T) {
    this.value = value;
    this.iterated = false;
  }

  hasNext(): boolean {
    return !this.iterated;
  }

  next(): T {
    if (this.iterated) {
      throw new Error("Iteration is over.");
    }
    this.iterated = true;
    return this.value;
  }
}

export class NestedIterator implements ListIterator<number> {
  private stack = new Array<ListIterator<number>>();

  constructor(nestedList: NestedInteger[] | NestedInteger) {
    if (Array.isArray(nestedList)) {
      this.initializeFromList(nestedList);
    } else if (nestedList.isInteger()) {
      this.stack.push(new Single<number>(nestedList.getInteger()!!));
    } else {
      this.initializeFromList(nestedList.getList());
    }
  }

  hasNext(): boolean {
    try {
      this.lastStackItem();
      return true;
    } catch (err) {
      return false;
    }
  }

  next(): number {
    return this.lastStackItem().next();
  }

  private initializeFromList(nestedList: NestedInteger[]): void {
    for (let i = nestedList.length - 1; i >= 0; i--) {
      this.stack.push(new NestedIterator(nestedList[i]));
    }
  }

  private lastStackItem(): ListIterator<number> {
    while (this.stack.length > 0 && !this.stack[this.stack.length - 1].hasNext()) {
      this.stack.pop();
    }
    if (this.stack.length === 0) {
      throw new Error("No elements found");
    }
    return this.stack[this.stack.length - 1];
  }
}

/**
 * Your ParkingSystem object will be instantiated and called as such:
 * var obj = new NestedIterator(nestedList)
 * var a: number[] = []
 * while (obj.hasNext()) a.push(obj.next());
 */
